dsh.1
.\"  -*- nroff -*-
.\"
.\" Author: Matthew T. Piotrowski <mtp22@cornell.edu>
.\" Copyright (c) 2001 Matthew T. Piotrowski <mtp22@cornell.edu>, 
.\"                    All rights reserved
.\"
.\" Help with nroff syntax from the OpenSSH manpage, v1.99
.\"
.\" This manpage can be used and distributed under the same terms as
 \" dsh itself.
.\"
.\" $Id$
.Dd July 10, 2001
.Dt DSH 1
.Sh NAME
.Nm dsh
.Nd the distributed shell
.Sh SYNOPSIS
.Nm dsh
.Op Fl a
.Op Fl e Ar 'command'
.Op Fl f
.Op Fl h
.Op Fl N Ar nodegroup1, nodegroup2, ...
.Op Fl w Ar node1, node2, ...
.Op Ar command
.Sh DESCRIPTION
.Nm
(the distributed shell) is a program which executes a single command
on multiple remote machines.  It can execute this command in parallel
(i.e. on any number of machines at a time) or in serial (by
specifying parallel execution of the command on 1 node at a time).
It was originally designed to work with rsh, but has full support for
ssh and with a little tweaking of the top part of
the script, should work with any program that allows remote execution
of a command without an interactive login.
.Pp
.Nm
takes a command and a set of nodes from the user; it checks the nodes
to make sure they can be contacted, and then runs the command on each node.
.Nm
displays the output (STDERR and STDOUT) of each
.Ar node
preceded by its name. The order the nodes are displayed is 
described in 
.Cm "A Brief Description of How dsh Removes Duplicates" ,
in the Section 
.Cm SPECIFYING NODES .
.Pp
.Nm
can also be run in interactive mode.  This mode is entered when the
user doesn't specify a command on the command line.  In this mode,
.Nm
will present the user with a prompt.  Here the user can enter a
command and 
.Nm
will run the command on the nodes that the user
specified when 
.Nm
was started.  After the command is run, the user
will again be presented with a prompt and can enter another command.  
It is important to note that this mode
does not actually log in to all of the nodes: it executes each command
separately using the remote execution program specified (typically
rsh or ssh).  An important consequence of this is that any properties
of a login shell, such as setting environment variables and job control, are 
not present.  Although
.Nm
does not actually log in to each node, commands entered at the 
.Nm
prompt are not interpreted by the local shell.  So entering 
echo $PATH at the 
.Nm
prompt 
.Pq Pa dsh> echo $PATH 
is different than entering echo
$PATH at the regular command line 
.Pq Pa dsh echo $PATH .  
In the first
case, the command echo $PATH is passed to the remote computer as is,
and so the PATH variable refers to the variable on
the remote machine.  In the second case, the variable PATH is first 
interpretted by the local shell, which replaces $PATH with the value
of PATH on the local machine before the command is passed to
the remote machine.  To simulate the first case at the regular
command line, use the -e switch 
.Pq Pa dsh -e 'echo $PATH' 
or quote the
variable to avoid interpolation by the local shell 
.Pq Pa dsh echo '$PATH' .
So, in short, commands typed at the
.Nm
prompt are executed as if for each node you typed 
.Pa rsh [remote computer] '[command you enter]' , 
where the command you enter is
placed in single quotes.  If you chose to add support for the
Term-ReadLine-Gnu Perl module during installation, then the 
.Nm
prompt has command recall (up and down arrows) and editing 
(as featured in the GNU readline library).  Also, a local command 
can be executed by placing an exclamation point (!) in front 
of the command (e.g.
.Pa dsh> !ls 
runs ls on the local computer).
.Sh SPECIFYING NODES
There are 4 ways to specify where
.Nm
should execute a command:
.Ss -a
This switch adds all the nodes in the file
$BEOWULF_ROOT/node_groups/ALL, where ALL is a file specifying one node
per line (IP address or hostname).  Lines that begin with a # and
lines that have only whitespace are ignored.

.Pa example:
.nf
	  mtp6@front_end $ cat /util/beowulf/node_groups/ALL

	  front_end
	  # Linux nodes
	  linux1
	  linux2
	  penguin1
	  penguin2
	  # OpenBSD nodes
	  openbsd1
	  openbsd2
	  openbsd3
	  # FreeBSD nodes
	  freebsd1
	  freebsd2

	  mtp6@front_end $ dsh -a ls

.fi
This will run the command ls on front_end, linux1, linux2,
penguin1, penguin2, openbsd1, openbsd2, openbsd3, freebsd1, and
freebsd2
.Ss -N nodegroup1, nodegroup2, nodegroup3, ...
This switch adds all the nodes in the file
.br
$BEOWULF_ROOT/node_groups/nodegroup1 and all the nodes in the file
.br
$BEOWULF_ROOT/node_groups/nodegroup2, etc.
As with the ALL file,
the format of these nodegroup files is one node per line (IP address
or hostname).  Lines that begin with a # and lines that have only 
whitespace are ignored.

.Pa example:
.nf
	  mtp6@front_end $ cat /util/beowulf/node_groups/linux

	  # Linux part of the cluster, running Slackware 8.0, kernel 2.4.6
	  linux1
	  linux2
	  penguin1
	  penguin2

	  mtp6@front_end $ cat /util/beowulf/node_groups/openbsd

	  # OpenBSD part of the cluster, running OpenBSD 2.9
	  openbsd1
	  openbsd2
	  openbsd3

	  mtp6@front_end $ dsh -N linux, openbsd ls

.fi
This will run the command ls on linux1, linux2,
penguin1, penguin2, openbsd1, openbsd2, and openbsd3
.Ss -w node1, node2, node3, ...
This switch adds the nodes node1, node2, node3, etc.

.Pa example:
.nf
	  mtp6@front_end $ dsh -w freebsd1, freebsd2 ls

.fi
This will run the command ls on freebsd1 and freebsd2

.Pa example:
.nf
	  mtp6@front_end $ dsh -w 10.10.10.1, 10.10.10.2, 10.10.10.3 ls

.fi
This will run the command ls on 10.10.10.1, 10.10.10.2, and 10.10.10.3
.Ss WCOLL
If none of the above switches (-a, -N, -w) is specified, dsh defaults
to adding the nodes specified in the file pointed to by WCOLL.
.Nm
only looks at this environment variable when none of the 3 switches
described above is specified.

.Pa example:
.nf
	  mtp6@front_end $ export WCOLL=/home/mtp6/bsd

	  mtp6@front_end $ cat /home/mtp6/bsd

	  openbsd1
	  openbsd2
	  openbsd3
	  freebsd1
	  freebsd2

	  mtp6@front_end $ dsh ls

.fi
This will run the command ls on openbsd1, openbsd2, openbsd3, freebsd1, and
freebsd2
.Ss Combining -a, -N, and -w
You can combine the -a, -N, and -w switches in any order.

.Pa example:
.nf
	  mtp6@front_end $ dsh -N linux -w front_end ls

.fi
This will run the command ls on linux1, linux2, penguin1, penguin2,
and front_end
.Ss Duplicate Nodes
.Nm 
has a somewhat complicated way of dealing with duplicate nodes
.Pp
For example, if the node which has IP address 10.10.10.3 is called
linux3 and is aliased to tux3, then entering 
.Pa dsh -w linux3, tux3 ls
will only execute the command on 10.10.10.3 once.
Similarly,
.Pa dsh -w linux3, 10.10.10.3 ls
will only execute the command on 10.10.10.3 once.
.Pp
A more complicated example is a computer which has more than one
IP address.  For example, say front_end has a local IP address
10.10.10.254 and a remote IP address 128.205.83.101.  The following
commands are equivalent in that they will execute the command ls on
front_end only once (the IP address used to contact this node is the
first IP address returned by gethostbyname(3)):

.nf
	dsh -w front_end ls
	dsh -w 10.10.10.254 ls
	dsh -w 128.205.83.101 ls
	dsh -w front_end, 10.10.10.254 ls
	dsh -w 10.10.10.254, front_end ls
	dsh -w front_end, 128.205.83.101 ls
	dsh -w 128.205.83.101, front_end ls
	dsh -w front_end, 10.10.10.254, 128.205.83.101 ls
	dsh -w front_end, 128.205.83.101, 10.10.10.254 ls
	dsh -w 10.10.10.254, front_end, 128.205.83.101 ls
	dsh -w 128.205.83.101, front_end, 10.10.10.254 ls

.fi
In general, dsh will remove duplicates; however, there are cases where
dsh won't remove duplicates because it cannot tell from IP
addresses alone whether they refer to the same computer (i.e. dsh
doesn't do reverse name lookups when looking for duplicates).  For 
example, in the above scenario, the combinations which will
result in
.Nm
executing the command more than once are those which start with
the two IP addresses:

.nf
	dsh -w 10.10.10.254, 128.205.83.101 ls
	dsh -w 128.205.83.101, 10.10.10.254 ls
	dsh -w 10.10.10.254, 128.205.83.101, front_end ls
	dsh -w 128.205.83.101, 10.10.10.254, front_end ls
.fi

To see why, refer to the next section.

.Ss A Brief Description of How dsh Removes Duplicates 
.Nm
looks at each node, in the order of -w nodes first, and -a and -N
nodegroups second (the order of these last two depending on the order
they appear on the command line).  For each node,
.Nm 
resolves the name
to a set of IP addresses using 
.Pa gethostbyname(3) ; 
if any of these IP
addresses has already been seen, the node is removed from the list
(note: in the case of duplicates found in nodes specified by -a and
-N, the nodes are not removed from the actual files, they are only
removed from the current list, which is stored in RAM).

.Pa example:
.nf
	  mtp6@front_end $ dsh -w front_end, 128.205.83.101 ls

.fi
This will run the command on front_end only once because
.Nm
performs the following duplication analysis:  It looks at front_end
and resolves it into the IP addresses 10.10.10.254 and
128.205.83.101.  Neither of these IP addresses have been seen before,
so it doesn't remove front_end from the list.
.Nm
then looks at 128.205.83.101 and resolves it into the IP address
128.205.83.101.  This address has been seen before, so it removes
128.205.83.101 from the list.

.Pa example:
.nf
	mtp6@front_end $ dsh -w 10.10.10.254, 128.205.83.101 ls

.fi
This will run the command on front_end twice because
.Nm
performs the following duplication analysis:  It looks at 10.10.10.254
and resolves it into the IP address 10.10.10.254.  This IP address
hasn't been seen before, so it doesn't remove 10.10.10.254 from the
list.
.Nm 
then looks at 128.205.83.101 and resolves it into the IP address
128.205.83.101.  This address hasn't been seen before, so it doesn't
remove 128.205.83.101 from the list.

If you're not sure about duplicates, you can use the -q switch to see 
the list of nodes where
.Nm
would execute the command without actually executing the command.
.Sh ENVIRONMENT VARIABLES
.Ss
.Ss BEOWULF_ROOT
This is the directory that 
.Nm
uses to search for node groups
(specified with the -N switch) and the ALL file (specified with the -a
switch).  Note that whatever directory is specified in the variable
BEOWULF_ROOT must have a subdirectory called node_groups where the
various nodegroup files and the ALL file are located.  
.br The default value of BEOWULF_ROOT is /util/beowulf
.Ss FANOUT
This is the number of nodes to run the command on in parallel.  The
default value is all of the nodes; however, you may wish to limit this
value if you have an extremely large number of nodes or if the machine
where you are running dsh from has a small amount of free resources.
For each node, 3 processes are forked, so to calculate the number of
processes running at one time, multiply FANOUT by 3 and add 1 (for
.Nm dsh )
.Ss WCOLL
See SPECIFYING NODE GROUPS
.Sh SWITCHES
.Bl -tag -width Ds
.It Fl N Ar nodegroup1, nodegroup2, ...
See SPECIFYING NODE GROUPS
.It Fl a
See SPECIFYING NODE GROUPS
.It Fl w Ar node1, node2, ...
See SPECIFYING NODE GROUPS
.It Fl q
Lists the nodes where dsh would execute the command without actually
executing the command
.It Fl e Ar 'command'
Executes the command in quotes on all the specified nodes.  The reason for this
switch is that putting the command to execute at the end of the
command line without single quotes can lead to metacharacter issues
(see 
.Cm QUOTING SHELL METACHARACTERS
below).  This way of specifying the commmand avoids this problem.
.It Fl t Ar (time in seconds)
Specifies the time to wait for a node to respond before labelling it
"unreachable".  The default value is 5.  Specifying 0 indicates no
timeout.
.It Fl f
If this flag is specified, 
.Nm
won't prompt the user whether or not to
continue if a node is unreachable or refusing a remote connection.
.Nm
will assume it should continue, bypassing the node. This is useful if
.Nm
is run in a non-interactive script.
.It Fl h
Displays a message containing a brief description of the command line
switches
.Sh QUOTING SHELL METACHARACTERS
When
.Nm
is used in the form
.Pa dsh [options] command ,
there is an issue with shell metacharacters.  That is, characters in
.Pa command 
are interpretted by the local shell before they are passed to
.Nm dsh .
For example, 
.Pa dsh -w 10.10.10.1, 10.10.10.2 find / -name 
.Pa bash > bash.find
will be interpretted as 'run dsh -w 10.10.10.1, 10.10.10.2 find /
-name bash and place the results of this program in the file
bash.find on the local computer'.  This does not run find / -name
bash on 10.10.10.1 and 10.10.10.2 and place the results of find 
in files named bash.find on the remote computers.  
To do that you would need to type
.Pa dsh -w 10.10.10.1, 10.10.10.2 find / -name 
.Pa bash '>' bash.find 
to avoid interpretation of > by the local shell.  Another way to avoid
local metacharacter interpretation is to use the -e switch: 
.Pa dsh -w 10.10.10.1, 10.10.10.2 -e 'find / -name 
.Pa bash > bash.find' .
I highly recommend using the -e switch to execute commands or running
dsh in interactive mode because most modern shells have many 
metacharacters and quoting these can be tricky even if you do 
recognize all of them.
.Sh EXIT CODES
If 
.Nm
exits cleanly (that is, if 
.Nm
is allowed to call exit itself),
the exit code of 
.Nm
will either be a 0 or a 1.  0 indicates that
.Nm
at least ran the command "rsh (node) command" for every non-duplicate
node.  Note that an exit code of 0 does not mean that every rsh was 
successful nor does it mean that the command exited with a code of 
0 on the remote computer :-\\ . The way 
.Nm
executes rsh commands doesn't lend itself to
capturing their exit codes.  An exit code of 1 indicates
that 
.Nm
did not attempt to run the command on any remote computers 
because of an error.
.Sh WEBSITE
http://dsh.sourceforge.net
.Sh AUTHORS
.Nm
was originally written by Jason Rappleye at the Center for
Computational Research at the University at Buffalo.  Version 2.0 was
written by Matthew T. Piotrowski at the Center for Computational
Research.  Thanks to dale@bewley.net for his contributions to the
early stages of version 2.0.  This manpage was written by Matthew
T. Piotrowski. The OpenSSH manpage (ssh.1), version 1.99, was used as an nroff
reference.
.Sh SEE ALSO
.Xr rsh 1 ,
.Xr gethostbyname 3 ,
.Xr ssh 1


CHANGELOG
Jul 26 2001 dsh.pl mtp : added a more detailed discussion of Term-ReadLine
	                 vs. Term-ReadLine-Gnu

Jul 26 2001 configure mtp : added the user option "--no-readline"

Jul 26 2001 configure mtp : dsh.1.gz to dsh.1, dsh*.tar.gz to dsh-*.tar.gz,
			    added backup port number in case getservbyname
	                    doesn't work

Jul 26 2001 TODO mtp : added this file

Jul 26 2001 README_SOLARIS mtp : added this file

Jul 26 2001 README_AIX mtp : added this file

Jul 26 2001 INSTALL mtp : referenced website, added GNU readline library info,
			  removed statement that Term-ReadLine-Gnu is included
			  with dsh

Jul 26 2001 CHANGELOG mtp : started this file


dsh.pl
#!/usr/bin/perl -w
use strict;
###############################################################################
# 
# dsh - distributed shell
#
# Authors : Jason Rappleye
#           Center for Computational Research at the University at Buffalo
#           rappleye@ccr.buffalo.edu
#
#           Matthew T. Piotrowski
#           Center for Computational Research at the University at Buffalo
#           mtp22@users.sourceforge.net
# 
# Last Modified: 07/17/2001
#
# Copyright (c) 2000-2001 State University of New York at Buffalo
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
# Modification history:
#
# 03/17/2000 jcr : initial release
#
# 07/17/2001 mtp : added parallel execution of the command, added
#    verification of the status of the nodes before execution of the command,
#    based the removal of duplicate nodes on IP addresses instead of hostnames,
#    added standard error to the node output, fixed the quoting of shell
#    metacharacters, fixed the formatting of the output when the names of nodes
#    are different lengths, added the -e switch, implemented fanout, added the
#    -t switch, added the -f switch, fixed -h switch, reorganized the code, 
#    and renamed some variables; also incorporated Dale's changes
#    
###############################################################################

# developer's note: if anything from here to the line of all #'s is edited,
# this change should also be reflected in the configuration script

use IO::Handle;
use Socket;
# if Term-ReadLine-Gnu is not installed, this call will not generate
# an error because the standard Perl distribution includes a Term-ReadLine
# module, which is what this call refers to.  This module, Term-ReadLine, is 
# mostly an interface to C readline libraries for other Perl modules (for 
# example, Term-ReadLine-Gnu); it doesn't implement many of the ReadLine 
# functions itself, but it does implement enough of these functions that 
# calling Term::ReadLine->new doesn't fail if Term-ReadLine-Gnu is not installed
use Term::ReadLine;

# User Configuration #######################################################
# the command that is used to execute your command (the one you specify 
# to dsh) on the nodes you specify to dsh (note: if you change this command,
# for example, to ssh, you need to change the following variable also;
# in this example, you would change it to getservbyname("ssh", "tcp")
my $RSH_CMD = "/usr/bin/rsh";
# the port used to contact the nodes you specify to dsh
# (note: getservbyname looks in the file /etc/services to determine
# the port number, so make sure there is a line in this file which has
# the service you specify) (also note: if you change this value to a
# non-standard port, for example, 91 for ssh, you need to edit the RSH_CMD
# variable so that it contacts the remote service on this port, for example,
# my $RSH_CMD = "/usr/bin/ssh -p 91")
my $RSH_CMD_PORT = getservbyname("shell", "tcp") || "514";
# directory where the "node_groups" folder is located
my $BEOWULF_ROOT = $ENV{BEOWULF_ROOT} || "/util/beowulf";
# name of file containing all nodes; located in $BEOWULF_ROOT/node_groups/
my $ALL_NODES = "ALL";
# number of nodes to process in parallel (this is approximately 1/3 of the 
# number of processes that will be running at one time); the default is 
# the total number of nodes
my $fanout = $ENV{FANOUT};
# the default time to wait for a node to respond when checking to see if 
# we can rsh to the node (note: this can also be set at the command line
# with the -t switch)
my $DEFAULT_TIMEOUT = 5;
############################################################################

my ($cmd, $cmdStart, $group, $i, $justList, $longest_hostname_length, $node,
    $timeout, $force,
    @cmd, @node_groups, @nodes,
    %hostname_length, %ip_addresses, %NODE_OUTPUT, %pid);

&process_command_line_switches;
&process_remaining_args;
&check_that_there_are_node_groups_or_nodes_to_work_with;
&expand_node_groups_to_nodes;
&remove_duplicate_nodes;
&check_that_we_can_rsh_to_nodes;
&find_length_of_longest_node_name;
&process_command;
exit;

sub process_command_line_switches {
    if ($#ARGV == -1 && ! defined $ENV{"WCOLL"}) {
	print STDERR "$0: no nodes specified, check the dsh man " .
	             "page for the 4 ways to specify nodes\n";
	&usage;
	exit(1);
    }
    ARG: for ($i = 0; $i <= $#ARGV; $i++) {
      SWITCH: {
	# need to break when $ARGV[$i] doesn't start with a '-'
	if ($ARGV[$i] !~ /^\-/) {
	  last ARG;
	}
	if ($ARGV[$i] eq "-N") {
	  do {
	      # check to see if nodegroup passed to the -N switch begins with
	      # a minus sign (-); if so, it is possible that this is
	      # an actual filename, but more likely it is an error
	      if ($ARGV[++$i] =~ /^\-/) {
		  print STDERR "$0: invalid nodegroup \"$ARGV[$i]\" specified ";
		  print STDERR "with -N switch\n";
		  &usage;
		  exit(1);
	      }
	      foreach (split(",", $ARGV[$i])) {
		  push @node_groups, "$BEOWULF_ROOT/node_groups/$_";
	      }
	  # allow for spaces in node group list 
          # (e.g. dsh -N nodegroup1, nodegroup2)
	  } while ($ARGV[$i] =~ /,$/);
      
	  last SWITCH;
	}
	if ($ARGV[$i] eq "-a") { 
	  push @node_groups, "$BEOWULF_ROOT/node_groups/$ALL_NODES"; 
	  last SWITCH; 
	}
	if ($ARGV[$i] eq "-w") {
	  do {
	      # check to see if the first node name passed to the -w switch
	      # begins with a minus sign (-); if so, it is possible that this is
	      # an actual hostname, but more likely it is an error
	      if ($ARGV[++$i] =~ /^\-/) {
		  print STDERR "$0: invalid node \"$ARGV[$i]\" specified ";
		  print STDERR "with -w switch\n";
		  &usage;
		  exit(1);
	      }
	      push @nodes, split(",", $ARGV[$i]);
	      # allow for spaces in node list 
	      # (e.g. dsh -w node1, node2)
	  } while ($ARGV[$i] =~ /,$/);

	  last SWITCH; 
	}
	if ($ARGV[$i] eq "-q") {
	  $justList = 1;
	  last SWITCH;
	}
	if ($ARGV[$i] eq "-e") {
	  # check to see if the command passed to the -e switch begins
	  # with a minus sign (-); if so, it is possible that this is
	  # an actual command, but more likely it is an error
	  if ($ARGV[++$i] =~ /^\-/) {
	      print STDERR "$0: invalid command \"$ARGV[$i]\" specified ";
	      print STDERR "with -e switch\n";
	      &usage;
	      exit(1);
	  }
	  $cmd = $ARGV[$i];
	  last SWITCH;
	}
	if ($ARGV[$i] eq "-t") {
	  if ($ARGV[++$i] !~ /^-?\d+$/) {
	      print STDERR "$0: invalid timeout \"$ARGV[$i]\" specified with " .
		           "-t switch\n";
	      &usage;
	      exit(1);
	  }
	  $timeout = $ARGV[$i];
	  last SWITCH;
	}
	if ($ARGV[$i] eq "-f") {
	  $force = 1;
	  last SWITCH;
	}
	if ($ARGV[$i] eq "-h") {
	  &usage;
	  exit;
	}
	print STDERR "$0: invalid switch \"$ARGV[$i]\":\n";
	&usage;
	exit(1);
      }
    }
}

sub process_remaining_args {
    # the remaining arguments are parts of the command to send to the nodes
    $cmdStart = $i;
    push @cmd, @ARGV[$cmdStart..$#ARGV];
    $cmd ||= join(" ", @cmd);  # command could already be defined by -e switch

    # quote the remaining arguments so that they are not interpretted again
    # by the local shell
    # (note:  this will also quote the command specified by the -e switch,
    #         if that switch was used)
    $cmd = "\'" . $cmd . "\'";
}

sub check_that_there_are_node_groups_or_nodes_to_work_with {
    # if there are no node_groups or nodes specified, check to see if a file was
    # specified in the environment variable "WCOLL"
    if ((! @node_groups && ! @nodes)) {
	if (defined($ENV{WCOLL})) {
	    push @node_groups, $ENV{"WCOLL"};
	}
	else {
	    print STDERR "$0: no nodes specified, check the dsh man " .
			 "page for the 4 ways to specify nodes\n";
	    exit(1);
	}
    }
}

sub expand_node_groups_to_nodes {
    # place the nodes in each file into the @nodes array 
    foreach $group (@node_groups) {
      open (FILE_IN, "< $group") or
	print STDERR "$0: couldn't open file \"$group\" : $!\n"
	    and exit(1);
      while (<FILE_IN>) {
	chomp;
	push @nodes, $_ unless (/^\#/ || /^\s*$/);
      }
      close FILE_IN;
    }
}

sub remove_duplicate_nodes {
    # remove duplicate nodes based on IP address
    my %ip_addresses_seen;
    @nodes = grep { my @name_lookup = gethostbyname($_)
			or print STDERR "$0: could not lookup " . 
			                "hostname \"$_\":  $!\n"
			     and exit(1);
		    # gethostbyname returns a couple different pieces of
		    # information; we are interested in the IP addresses
		    @name_lookup = @name_lookup[4 .. $#name_lookup];
		    my @resolved_ips = map {inet_ntoa($_)} @name_lookup;
		    # store one of the IP addresses so that $RSH_CMD
		    # doesn't have to translate it again later
		    $ip_addresses{$_} = $resolved_ips[0];
		    # check each resolved IP address to see if we have
		    # seen this node before; if we have seen any of
		    # the node's IP addresses before (either directly or
		    # by resolving an alias), we consider the node a duplicate
		    # and remove it from the list of nodes
		    my $have_seen_this_node = 0;
		    foreach my $ip (@resolved_ips) {
		       $have_seen_this_node = $ip_addresses_seen{$ip}++ ||
			                      $have_seen_this_node;
		    }
		    $have_seen_this_node == 0; } @nodes;
}

sub check_that_we_can_rsh_to_nodes {
    my $current_node_index = 0;
    while ($current_node_index < @nodes) {
	unless (we_can_rsh_to($nodes[$current_node_index])) {
	    unless ($force) {
		# couldn't rsh to node; ask user if we should continue
		# the program
		my $answer = "";
		until ($answer =~ /^y$/i || $answer =~ /^n$/i) {
		    print "The command hasn't been run on any nodes yet, " .
			  "would you like to continue?\n" .
			  "[y or n] ";
		    $answer = <STDIN>;
		    chomp($answer);
		}
		exit(1) unless ($answer =~ /^y$/i);
	    }
	    # remove node from node array
	    splice(@nodes, $current_node_index, 1);
	    # don't increment the current node index: the next node is actually
	    # at the current index because we removed an element from the array
	    # (note: this is the reason for using a while loop instead of a
	    #        foreach)
	    next;
	}
	$current_node_index++;
    }
}

sub we_can_rsh_to {

    my $node = shift;

    my $node_ip_address = $ip_addresses{$node};
    my $port = $RSH_CMD_PORT;  # standard rsh port
    my $socket_structure = sockaddr_in($port, inet_aton($node_ip_address));

    # make the socket
    my $proto = getprotobyname('tcp');
    socket(SOCKET_HANDLE, PF_INET, SOCK_STREAM, $proto);

    # install timeout alarm
    local $SIG{ALRM} = sub {
	                     print STDERR "$0: couldn't connect to $node: \n" .
				          "$0: It appears the node is down " . 
					  "or unreachable.\n";
                             die "connection timed out\n";
               	       };
    # set the default timeout if timeout is not already defined
    $timeout = $DEFAULT_TIMEOUT unless defined($timeout);
    if ($timeout < 0) {
	print STDERR "$0: $timeout is not a valid timeout value.\n";
	print STDERR "$0: using default timeout value ($DEFAULT_TIMEOUT)\n";
	$timeout = $DEFAULT_TIMEOUT;
    }
    alarm($timeout);
    eval {
	unless (defined (connect(SOCKET_HANDLE, $socket_structure))) {
	  print STDERR "$0: couldn't connect to $node: \n" .
                       "$0: It appears the node is up but isn't " .
                       "allowing incoming rsh connections.\n";
	  die "connection refused\n";
	}
    };
    close(SOCKET_HANDLE);
    alarm(0);
    if ($@) {
	# couldn't connect to node
	return undef;
    }
    else {
	# connected to node successfully
	return 1;
    }
}

# find the length of the longest node name to properly format the node output
sub find_length_of_longest_node_name {
    $longest_hostname_length = 0;
    foreach $node(@nodes) {
	my $current_hostname_length = length $node;
	if ($current_hostname_length > $longest_hostname_length) {
	    $longest_hostname_length = $current_hostname_length;
	}
	$hostname_length{$node} = $current_hostname_length;
    }
}

sub process_command {
    if ($justList) {
      print join("\n", @nodes), "\n";
      exit(0);
    }

    if ($cmd && ($cmd ne '\'\'')) {
	&run_cmd_in_parallel();
    }
    else {
	print << 'INTERACTIVE_DSH';
--------------------------------interactive dsh---------------------------------
- This prompt features command history (using the up and down arrows), command
  completion (relative to the local computer), and command editing (for example,
  the home and end keys).
- Commands preceded by an exclamation point (!) will be run on the local
  computer. Note that these local commands will have "2>&1" appended to 
  merge standard error and standard output
--------------------------------------------------------------------------------
INTERACTIVE_DSH
      my $term = Term::ReadLine->new("DSH");
      while ($cmd = $term->readline("dsh> ")) {
	# check to see if the command is a local command
	if ($cmd =~ s/^!//) {
	  print "executing $cmd\n";
          my $LOCAL_OUTPUT = new IO::Handle;
          if (my $pid = open($LOCAL_OUTPUT, "-|")) {}
	  elsif (defined $pid) {
	      exec "$cmd 2>&1"
		  or print "couldn't run command '$cmd': $!\n"
		      and exit;
	  }
	  else {
	      print STDERR "$0: can't fork additional process to " .
		           "run command '$cmd'\n";
	  }
          while (<$LOCAL_OUTPUT>) {
	      print "local computer: \t", $_;
	  }
	  close($LOCAL_OUTPUT);
	  # flush STDOUT
	  $| = 1;
	  $| = 0;
	}
	# check to see if the user entered exit
	elsif ($cmd =~ m/^\s*exit\s*$/) {
	    exit;
	}
	else {
	  $cmd = "\'" . $cmd . "\'";
	  &run_cmd_in_parallel();
	}
      }
    }
}

sub run_cmd_in_parallel {
  my $total_number_of_nodes = @nodes;
  # number of nodes to rsh to in parallel; default is the total number of nodes
  $fanout = $total_number_of_nodes unless defined($fanout);
  if ($fanout < 1) {
     print STDERR "$0: $fanout is not a valid fanout value:\n";
     print STDERR "$0: using closest valid fanout value (1)\n";
     $fanout = 1;
  }
  if ($fanout > $total_number_of_nodes) {
     print STDERR "$0: the fanout value specified ($fanout) is larger " .
	          "than the number of nodes:\n";
     print STDERR "$0: using closest valid fanout value " .
	          "($total_number_of_nodes)\n";
     $fanout = $total_number_of_nodes;
  }
  print "executing $cmd\n";
  # flush STDOUT before forking to avoid duplicate output when STDOUT is
  # in block-buffered mode (for example, when the standard output of dsh is
  # redirected to a file)
  select STDOUT;
  $| = 1;
  $| = 0;
  # I haven't witnessed the same problem with STDERR as I have with STDOUT, but
  # to be safe, I am also flushing STDERR 
  select STDERR;
  $| = 1;
  $| = 0;
  for (my $starting_node = 0; 
       $starting_node < $total_number_of_nodes; 
       $starting_node += $fanout) {
    my $ending_node = $starting_node + $fanout - 1;
    if ($ending_node >= $total_number_of_nodes) {
	$ending_node = $#nodes;
    }
    foreach my $node (@nodes[$starting_node..$ending_node]) {
      FORK: {
	  $NODE_OUTPUT{$node} = new IO::Handle;
	  if ($pid{$node} = open($NODE_OUTPUT{$node}, "-|")) {}
	  elsif (defined $pid{$node}) {
	      exec "$RSH_CMD $ip_addresses{$node} $cmd 2>&1" 
		  # note: 2>&1 merges standard error with standard output
		  or print "couldn't $RSH_CMD to this node: $!\n"
		      and exit;
	  }
	  else {
	      print STDERR "$0: can't fork additional process to " .
		           "$RSH_CMD to $node.\n";
	      print STDERR "$0: command not run on $node: $!\n";
	  }
      }
    }
    # print node output to the screen in the same order as the @nodes array 
    foreach $node (@nodes[$starting_node..$ending_node]) {
	my $print_padding = 
	    " " x ($longest_hostname_length - $hostname_length{$node});
	# dereference hash before reading from file handle
	my $NODE_OUTPUT = $NODE_OUTPUT{$node};
	while (<$NODE_OUTPUT>) {
	    print "$node:$print_padding \t", $_;
	}
	close($NODE_OUTPUT);
    }
  }
}

sub usage {
    print << "USAGE";
usage: $0
    -a adds all nodes in the file \$BEOWULF_ROOT/node_groups/ALL
    -e 'command'
       executes the command on all the nodes
    -f if this flag is specified, dsh won't prompt the user whether or not to
       continue if a node is unreachable or refusing a remote connection
    -h display this information
    -N group1, group2, ... 
       adds the nodes in the files \$BEOWULF_ROOT/node_groups/group1,
       \$BEOWULF_ROOT/node_groups/group2, etc.
    -q lists the nodes where dsh would execute the command without actually
       executing the command
    -t time_in_seconds 
       specifies the time to wait for a node to respond before labelling it 
       "unreachable"
    -w node1, node2, ... 
       adds the nodes node1, node2, etc.
USAGE
}
